dough-inventories/src/main/java/io/github/bakedlibs/dough/inventory/SlotGroupIterator.java

```diff
@@ -0,0 +1,51 @@
+package io.github.bakedlibs.dough.inventory;
+
+import java.util.Iterator;
+import java.util.NoSuchElementException;
+
+import javax.annotation.Nonnull;
+
+/**
+ * This {@link Iterator} implementation iterates through all slots within a {@link SlotGroup}.
+ * 
+ * @author TheBusyBiscuit
+ *
+ */
+class SlotGroupIterator implements Iterator<Integer> {
+
+    private final int[] slots;
+    private int index = 0;
+
+    /**
+     * This creates a new {@link SlotGroupIterator} for the given
+     * {@link SlotGroup}.
+     * 
+     * @param slotGroup
+     *            The {@link SlotGroup}
+     */
+    SlotGroupIterator(@Nonnull SlotGroup slotGroup) {
+        this.slots = slotGroup.getSlots();
+    }
+
+    @Override
+    public boolean hasNext() {
+        return index < slots.length;
+    }
+
+    @Override
+    public Integer next() {
+        if (index < slots.length) {
+            int slot = slots[index];
```

Comment:
can't this be inlined?
`int slot = slots[index++];`

Reply:
inline de/increments should be avoided where possible.

They violate a core principle regarding Java's order of evaluation.
This commonly used bad design pattern is tracked here:
https://wiki.sei.cmu.edu/confluence/display/java/EXP05-J.+Do+not+follow+a+write+by+a+subsequent+write+or+read+of+the+same+object+within+an+expression

Reply:
I see. Noted for the future.